### PR TITLE
USB: Fix reconnect issues

### LINKED ISSUE #15917: USB disconnect/reconnect
Hardware/Software
nrf52840 on Nordic PCA100056 DK
USB configured for either CDC-ECM or CDC_ACM

** Describe the bug **
Detaching the USB cable from the host and reattaching (not timing dependent) will not reinitialize/reenumerate the device correctly.

**To Reproduce**
1. build the cdc_acm sample for your board
2. flash and run on hardware
3. verify that USB device enumerates property on initial attach (or attached at boot)
4. pull USB cable from host, reattach after some period of time
5. verify that the USB device no longer words or enumerates on the host.

**Expected behavior**
The USB device should re-enumerate correctly and become functional after the reconnect.

**Impact**
Stability of product.  This would affect almost all users of USB.

**Screenshots or console output**

**Environment (please complete the following information):**
 - OS: Linux (ubuntu 16.04 and/or 18.04
 - Toolchain Zephyr SDK, 
 

**Additional context**
I have made a few changes to the code to provide a work-around/fix, however I believe the biggest issue is the architecture of the stack. usb structures do not get properly shutdown or disabled upon usb detach.
The first part of my fix, is to clear the context enabled flag for each end-point in ep_ctx_reset. It was found that upon reenumeration, the function usb_dc_ep_enable was unable to enable the endpoint because this flag was set. This function then returned an error code that was unexpected by the caller, which made the caller fail.
Fix 1 allows the usb device to reenumerate correctly, however, transaction records are lost when this happens (because transactions are pending at time of reconnect).  Eventually, enough transactions are lost that the usb_device code can no longer start transactions and reenumeration will again start failing (multiple detach/reattach attempts must be made).
Fix part 2. This fix includes a loop that will cancel all pending transactions on USB_DC_RESET messages.  This allows the kernel to free the semaphores that have been locked on pending transactions.  This also works, but then enters another issue.
Issue 3.  During the USB_DC_RESET sequence, higher level parts of the stack have not been informed of the condition of the USB bus, and try to continue sending/receiving USB transactions over interface endpoints (non-control endpoints).  These new transaction requests can come in at any time, even after cancelling transactions.  Again, these transactions are lost by the system, and semaphores remain locked, eventually allowing the usb_device to run out of usb transaction slots and incapable of starting new transactions.
Fix 3.  A new BOOL has been added to the usb_dev_priv local structure that gets cleared on USB_DC_RESET event, and set only after a USB_DC_INTERFACE event.  The function usb_transfer() has been modified to check this flag before starting a new transaction and if the new transaction is destined for any endpoint other than the control endpoint ((ep & 0x7F)==0) the function will return an error to the caller and the transaction not started.

These 3 fixes now make my devices and hardware stable on USB disconnect and reconnects. The total changes amount to about 15 lines of code in usb_device.c and usb_dc_nrfx.c.

### PR DESCRIPTION
**Pull request by finikorg**
I still need to check suspend / resume properly working.

Fixes #15917

### CODE REVIEW

File: subsys/usb/usb_device.c [hunk 1]
```diff
@@ -937,10 +943,51 @@ static void usb_register_status_callback(usb_dc_status_callback cb)
...
+	}
+
+	return 0;
+}
+
+static int disable_interface_ep(const struct usb_ep_cfg_data *ep_data)
+{
+	u8_t ep = ep_data->ep_addr;
+
+	if (ep | !0x80) {
```
Comment (rakons) [inline subsys/usb/usb_device.c]: What it is supposed to do?

**Review by rakons — CHANGES_REQUESTED** [review]

File: subsys/usb/usb_device.c (hunk 1 above)
Comment (finikorg) [reply subsys/usb/usb_device.c]: Sorry, my bad, removed that

**Review by loicpoulain — COMMENTED** [review]
LGTM

File: subsys/usb/usb_device.c (hunk 1 above)
Comment (rakons) [reply subsys/usb/usb_device.c]: In general, it is good not to allow to disable EP0 and I suppose it was relating to that issue. I am not sure if the underlying library does take care of this?

Comment (finikorg) [reply subsys/usb/usb_device.c]: That endpoint table contains only non control endpoints, so we do not need this (incorrect) check.

**Review by barsok — APPROVED** [review]
Found an issue which rakon then reported -- fixed. Verified on an own piece of HW.

**Review by masz-nordic — CHANGES_REQUESTED** [review]
This allows USB3CV tests to be executed again, but following cases from Chapter 9 Tests still fail:

- TD 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Configured state: 
  - `Set Configuration failed for Configuration Value : 1`
  - `(1.1.2) Devices must support being set to Addressed/Configured state.`

- TD 9.9, 9.13: 
  - `Set Configuration failed for Configuration Value : 1`
  - `(1.1.2) Devices must support being set to Addressed/Configured state.`

- TD 9.16:
  - `Enumeration failed at iteration 1 (General failure)`

**Comment by finikorg** [comment]
> This allows USB3CV tests to be executed again, but following cases from Chapter 9 Tests still fail:
> 
>     * TD 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Configured state:
>       
>       * `Set Configuration failed for Configuration Value : 1`
>       * `(1.1.2) Devices must support being set to Addressed/Configured state.`
> 
>     * TD 9.9, 9.13:
>       
>       * `Set Configuration failed for Configuration Value : 1`
>       * `(1.1.2) Devices must support being set to Addressed/Configured state.`
> 
>     * TD 9.16:
>       
>       * `Enumeration failed at iteration 1 (General failure)`

Are those failures resulting from the PR? Can they be reproduced on master?

**Comment by masz-nordic** [comment]
On master the USB3CV tests do not work at all:

```
SetConfiguration() failed.
  No device:  deleting node!
  Device configuration failed!
  SetConfiguration() failed.
  No device:  deleting node!
  Device configuration failed!
A device was not found.  
Check that device is plugged into correct port.
Couldn't initialize common subsystems.
Function CVCommon_InitializeTestSuite did not succeed.  Return value = 0x80000000
```

**Comment by finikorg** [comment]
> On master the USB3CV tests do not work at all:
> 
> ```
> SetConfiguration() failed.
>   No device:  deleting node!
>   Device configuration failed!
>   SetConfiguration() failed.
>   No device:  deleting node!
>   Device configuration failed!
> A device was not found.  
> Check that device is plugged into correct port.
> Couldn't initialize common subsystems.
> Function CVCommon_InitializeTestSuite did not succeed.  Return value = 0x80000000
> ```

OK, This is probably because of extra checks we have added. I think we should address this in a separate issue.

**Comment by finikorg** [comment]
@masz-nordic Does this fix the issues?
```
diff --git a/subsys/usb/usb_device.c b/subsys/usb/usb_device.c
index 9e70966d10..765766cd5d 100644
--- a/subsys/usb/usb_device.c
+++ b/subsys/usb/usb_device.c
@@ -500,7 +500,6 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
 
        if (usb_dc_ep_enable(ep_cfg.ep_addr) < 0) {
                LOG_ERR("Failed to enable endpoint %x", ep_cfg.ep_addr);
-               return false;
        }
 
        usb_dev.configured = true;
```

**Comment by masz-nordic** [comment]
@finikorg It fixes 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Configured state and 9.9. But 9.13 and 9.16 still fail.
Additionally, there are new failures:
- 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Default state
- 9.1, 9.2, 9.3, 9.4, 9.5, 9.7 in Addressed state

All of them fail with:
```
Set Configuration failed for Configuration Value : 0
(1.1.2) Devices must support being set to Addressed/Configured state.
```

**Comment by finikorg** [comment]
Then we need to "fix" another bug :)
```
        if (usb_dc_ep_configure(&ep_cfg) < 0) {
                LOG_ERR("Failed to configure endpoint %x", ep_cfg.ep_addr);
-               return false;
        }
```

**Comment by masz-nordic** [comment]
Still same failures as in previous case. Logs:
```
[00:01:11.285,858] <dbg> usb_hid.hid_custom_handle_req: Standard request: bRequest 0x9 bmRequestType 0x0 len 0
[00:01:11.285,858] <dbg> usb_device.usb_handle_std_device_req: REQ_SET_CONFIGURATION, conf 0x0
[00:01:11.285,858] <dbg> usb_device.usb_set_configuration: Device not configured - invalid configuration
[00:01:11.285,888] <dbg> usb_hid.hid_status_cb: cfg 0x200033a8 status 3
[00:01:11.285,888] <dbg> usb_hid.hid_do_status_cb: USB device configured
[00:01:11.285,888] <dbg> usb_device.usb_handle_std_device_req: USB Set Configuration failed
[00:01:11.285,888] <dbg> usb_device.usb_handle_request: Handler Error 0
[00:01:11.285,919] <dbg> usb_device.usb_print_setup: Setup: 0 9 0 0 0
[00:01:11.285,919] <dbg> usb_device.usb_handle_control_transfer: usb_handle_request failed
```

**Comment by finikorg** [comment]
@masz-nordic what sample are you using? I had tests with frdm_k64f and HID it was passing all tests.

**Comment by masz-nordic** [comment]
@finikorg `hid-mouse` with `nrf52840_pca10056`.

**Review by jfischer-no — APPROVED** [review]

**Comment by jfischer-phytec-iot** [comment]
Can you please apply and test again? (https://gist.github.com/acsplayon/b594d6b2207f8923edb071fdee499b36)

```diff
diff --git a/drivers/usb/device/usb_dc_nrfx.c b/drivers/usb/device/usb_dc_nrfx.c
index e3ba244b9c..86c7e39351 100644
--- a/drivers/usb/device/usb_dc_nrfx.c
+++ b/drivers/usb/device/usb_dc_nrfx.c
@@ -637,6 +637,8 @@ static void ep_ctx_reset(struct nrf_usbd_ep_ctx *ep_ctx)
        ep_ctx->buf.curr = ep_ctx->buf.data;
        ep_ctx->buf.len  = 0U;
 
+       ep_ctx->cfg.en = false;
+
        ep_ctx->read_complete = true;
        ep_ctx->read_pending = false;
        ep_ctx->write_in_progress = false;

```

**Comment by masz-nordic** [comment]
@jfischer-phytec-iot This fixes 9.16. Other still fail, here is the trace from Default state tests:



If this is nRF specific issue and this PR fixes USB3CV for other platforms, let's merge it and I will investigate fails in a separate case.

**Comment by finikorg** [comment]
I have tested frdm_k64f, stm32f723e_disco, nucleo_f412zg they are all passing, I will check also reel_board.

**Review by jfischer-no — APPROVED** [review]
LGTM

**Comment by finikorg** [comment]
@masz-nordic Tests are  passing for `reel_board` also if I disable `USB_DEVICE_REMOTE_WAKEUP`

**Comment by masz-nordic** [comment]
@finikorg: Which test mode are you using? I have tested `reel_board` with Compliance Test on Chapter 9 Tests [USB 2 devices] and it still fails.

**Comment by finikorg** [comment]
> @finikorg: Which test mode are you using? I have tested `reel_board` with Compliance Test on Chapter 9 Tests [USB 2 devices] and it still fails.

Yes Compliance "Chapter 9 Tests [USB 2 devices]", only Remote Wakeup TD 9.12 is failing.

I am on top of my above mentioned PR, my reel_board is the oldest version, probably you have newer one?

**Comment by jfischer-phytec-iot** [comment]
@masz-nordic Anything against to merge it?

**Comment by masz-nordic** [comment]
@finikorg Mine is `1507.1`. I disabled remote wakeup in `Kconfig` (apart from `prj.conf`) too and now it only fails 9.21 and 9.16.

@jfischer-phytec-iot No, if this fixes problems from the original issue it is fine. I will create a separate one for USB3CV testcases.

**Review by masz-nordic — APPROVED** [review]

**Review by rakons — APPROVED** [review]

**Pull request merged**